### ISSUE TITLE
BUG: Fusion parser may return wrong result for multi site project in development context

### ISSUE THREAD

**Issue by gjwnc** [neos/neos-development-collection#3835]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Prerequsites:
* Neos 8.0
* Multi site project
* Multiple (or all) site packages have Neos.Neos.fusion.autoInclude activated.
* Development context

To render a node the Neos.Neos FusionView tries to create a fusion runtime. To create the runtime, we need the runtimeconfiguration, thus we get to [FusionService::getMergedFusionObjectTree](https://github.com/neos/neos-development-collection/blob/master/Neos.Neos/Classes/Domain/Service/FusionService.php#L136)

The problem arises at the second request if it is another sitepackage than the first request, where the second request's sitepackage has autoInclude activated. This causes an exception with the message `No fusion object found in path "root"`.  (See below `Steps to reproduce` to understand what is happening)

### Why is this the case?
In `->parse()` the parserCache kicks in and checks if a cache entry exists for `$siteRootFusionPathAndFilename`. That is, in for the second request e,g, `resource://SitePackage2/Private/Fusion/Root.fusion`. This had already been auto included and thus already been parsed and cached. The cache returns this result and ignores `$mergedFusionCode` which would include all the other - valid - includes.

So, the core problem here is, that in development context, and activated site package autoinclude configuration, the parser cache returns an invalid result compared to 7.3. In 7.3 the project is working fine currently.

### Discussion

This specific behaviour raises a bunch of questions on how to fix this. Some of those questions are:
1. Is this a bug? Or just undocumented behaviour? On one hand - it was working until 8.0. On the other hand the [documentation](https://docs.neos.io/cms/manual/rendering/fusion#automatic-fusion-file-inclusion) "just" mentions that site package Root.fusion is loaded automatically. It doesn't mention that you should not autoinclude site packages and maybe there is a valid use case for auto loading site packages. I assume in our case it was just copy paste of some package :see_no_evil: 
If this is fixed in the documentation, there should be an exception in the case of autoloading of any site package - which might not be that easy to detect which of the packages are actually site packages?
2. I think the current behaviour just makes a corner case more obvious and `FusionService::getMergedFusionObjectTree` should make use of the parser's option to not provide `$siteRootFusionPathAndFilename`, e.g. `$this->fusionParser->parse($mergedFusionCode)`. This is an idea from @mhsdesign which we discussed yesterday. Are there any further issues with Neos.Fusion FusionView vs. Neos.Neos FusionView?
3. It is also a valid assumption, that the parser cache should not decide which code to parse, but just parse what was provided. E.g. in non Neos projects I would assume, that I can just wildly provide fusion code and it just parses it. Which it actually does in production, because the cache is disabled. So, is there a problem in the parsercache or just outside of the parser on how it is used?
4. Finally how should this be fixed?

### Quick fix
* Fix 1: Set `Neos.Fusion.enableParsePartialsCache: false` in development context
* Fix 2: Change site package configuration such that sitepackages are not autoincluded

### Expected Behavior

The second site should render as it was in 7.3 :lau

### Steps To Reproduce

Let's examine both requests in order to understand what is happening:
1. In `FusionService::getMergedFusionObjectTree` the source code is read from the corresponding Root.fusion site package, e.g. `$siteRootFusionPathAndFilename = 'resource://SitePackage1/Private/Fusion/Root.fusion';`. Now `$siteRootFusionCode` contains the file content of `$siteRootFusionPathAndFilename`.
2. The autoinclude's are added to the source code string. Because of site package autoinclude, in `$mergedFusionCode` we have the include for the second site package `include: resource://SitePackage2/Private/Fusion/Root.fusion`.
3. For the call to `$this->fusionParser->parse($mergedFusionCode, $siteRootFusionPathAndFilename)` it should be noted, that `$mergedFusionCode != \file_get_contents($siteRootFusionPathAndFilename)`. The `->parse()` call does it's job correctly and returns the correct array with many prototpyes found because of the autoincludes. 

The page renders correctly and the whole site 1 works fine.

4. The second request to site 2 begins and the process starts again in `FusionService::getMergedFusionObjectTree`. Autoincludes are added to `$mergedFusionCode`.
5. Now `$this->fusionParser->parse($mergedFusionCode, $siteRootFusionPathAndFilename)` returns an array without any autoincludes added, because `$siteRootFusionPathAndFilename` was already cached in the previous request and thus the rendering might fail with `No fusion object found in path "root"`.

### Environment

```markdown
- Flow: 8.0.4
- Neos: 8.0.3
- PHP: 8.1.4
```

**Comment by mhsdesign** [comment]
Thank you, @gjwnc for debugging this one - and sorry for the mess ;)

As discussed (and you pointed out a bit) there are a few ways to fix it:

## The strict way
1. If we feel like this issue arises because the `$contextPathAndFilename` should yield the `$sourceCode` - and `$contextPathAndFilename` must be null or a valid file name.
Then we can think of making it a requirement, to not provide `$contextPathAndFilename` in `FusionService->getMergedFusionObjectTree` if it differs. Also, the Fusionparser will throw an exception when `$contextPathAndFilename` is invalid: https://github.com/neos/neos-development-collection/pull/3834
```php
# ... $mergedFusionCode .=  ...
if (is_file($siteRootFusionPathAndFilename)) {
    $mergedFusionCode .= $this->getFusionIncludes([$siteRootFusionPathAndFilename]);
}
$mergedFusionCode .= $this->getFusionIncludes($this->appendFusionIncludes);

return $this->fusionParser->parse($mergedFusionCode);
```

## The lax legacy pre 8.0 way
2. If the behavior should be that Fusion should mostly care about `$sourceCode`, and the `$contextPathAndFilename` is allowed to differ from the `$sourceCode` and  `$contextPathAndFilename` is allowed to even not be an existing file at all https://github.com/neos/neos-development-collection/pull/3834
then we can say that the cache should not get confused by the manual creation of the `$sourceCode`. We can ensure that by not only saving the Ast in the cache identified by file name, but also add the md5 of the file contents. Later when retrieving from the cache we can check if the md5 of the content equals the cached - otherwise we know something is off and bypass the cache.

## The cleaner, strict way. 
3. Like 'The strict way', but we interact with the $fusionParser in a cleaner way.
Instead of creating a fusion file via string concatenation - we can add a way to use the parser by passing an array of includes:
```php
$fusionparser->parseIncludes([
  'AutoIncluded1.fusion',
  'SitePackageFile.fusion'
]);
```
But this doesnt seem to be an option as @kitsunet pointed out in https://neos-project.slack.com/archives/C04PYL8H3/p1657564600872759 that we need to keep the current `generateNodeTypeDefinitions` generator - which relies on creating fusion contents dynamically.

**Comment by kitsunet** [comment]
As said, we might need to extract that part from where it is right now, I wonder if we could cache it and provide that as some kind of special include (with some form of stream wrapper).

More and more I come to the conclusion that we (probably over time) diluted `\Neos\Fusion\Core\Parser::parse` and use the `contextPathAndFilename` for more than we should. I think it would be best from a "lets to it better" perspepctive to make arguments more explicit and separate the "context" aka the path for relative includes (which then could actually be a path and not a file) from the cache identifier to be used for the parsed results. Note that if we want to meaningfully cache we need to rethink that part as well if we explode includes out. I like the idea of giving includes specifically in some way, but then do we parse and cache them separately, or not at all (probably bad idea) or provide them to parse to cache it all together (how to generate a sensible cache identifier then)..... 

Many questions, seems like a topic to take a moment to discuss together with code at hand and try to find some sensible solutions together. I think it would be nice to envision best possible code with what we know now, and then see how much of that we can make happepn without breaking everything.

**Comment by gjwnc** [comment]
Regarding my own confusion for `a valid use case for auto loading site packages`: We use a custom package to get an overview of multiple elements of various sites, like a visual unit test which also depends on the autoload feature for the sites.  
Additionally, I think Sitegeist.Monocle depends also on this to render various prototypes.

**Comment by kitsunet** [comment]
Yeah I am not super happy about those cases but I also think it should somehow made to work, that it doesn't atm is really not founded in any sensible principle.

**Comment by kitsunet** [comment]
but magically deciding on an identifier is also not super great?

**Comment by kitsunet** [comment]
although I wonder we could maybe do the following:

`UnparseFusionCode` value object to which you add code in the right order via fromString or fromFile, no identiifers just the string or the file path, and it can calculate a cache identiifer based on a sha1 over the filepaths + strings or something. and we just pass that whole object to the fusion parser, which can then retrieve the appended code to parse (?) and the cache identifier? that might actually be a nice API...

**Comment by mhsdesign** [comment]
nope hehe as we see ^^ - but this is imo still the best / only choice we have. (with the assumption that we need the partial ast cache - which we do, to speed things up - but in the ideal world we wouldnt of course - no one likes a cache xD)

I spend a lot of thought into how the cacheIdentifier can be calculated and used and that is the way

----

i think it is just fatal to give an $contextPathAndFileName which doesnt return the contents of $sourceCode

**Comment by mhsdesign** [comment]
yes i like this value object, this will prevent doing bad stuff with the $contextPathAndFileName ;)

**Comment by kitsunet** [comment]
So to extend on that idea with pseudo code:

```
$unparsedFusionCodeCollection = new UnparsedFusionCodeCollection([
  UnparsedFusionCode::fromString($fusionCode), 
  UnparsedFusionCode::fromFile($filePath)
]);
$fusionParser->parse($unparsedFusionCodeCollection);
```

**Comment by mhsdesign** [comment]
yesss thats definitely cool ^^ thought about something similar so i guess we have a winner?

**Comment by mhsdesign** [comment]
i guess this still counts as bugfix right? so we could still add this to 8.0 right? - will prepare a PR

**Comment by mhsdesign** [comment]
https://github.com/neos/neos-development-collection/blob/b91861896db98628d435e399e1144919cf77404b/Neos.Fusion/Classes/View/FusionView.php#L194

should be adjusted, to use the new api too ;)
